### ISSUE TITLE
HMM sampling for out-of-sample data bug

### ISSUE THREAD

**Issue by thempel** [deeptime-ml/deeptime#149]
I am applying a learnt HMM to a different data set and want to extract representative structures. However, certain observable states are not in this data set (i.e. not in the discrete trajectories that I'm giving to this function). So this code will not work: 

```python
hmm.sample_by_observation_probabilities(dtrajs_off_sample, 50)

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-181-fb63165c514a> in <module>
----> 1 hmm.sample_by_observation_probabilities(dtrajs_off_sample, 50)
      2 observable_state_indices = sample.compute_index_states(mapped)

/storage/mi/thempel/anaconda3/envs/py37_feb21/lib/python3.8/site-packages/deeptime/markov/hmm/_hidden_markov_model.py in sample_by_observation_probabilities(self, dtrajs, nsample)
    542         mapped = self.transform_discrete_trajectories_to_observed_symbols(dtrajs)
    543         observable_state_indices = sample.compute_index_states(mapped)
--> 544         return sample.indices_by_distribution(observable_state_indices, self.output_probabilities, nsample)
    545 
    546     # ================================================================================================================

/storage/mi/thempel/anaconda3/envs/py37_feb21/lib/python3.8/site-packages/deeptime/markov/_sample.py in indices_by_distribution(indices, distributions, nsample)
    146     for dist in distributions:
    147         if len(dist) != n:
--> 148             raise ValueError('Size error: Distributions must all be of length n (number of states).')
    149 
    150     # list of states

ValueError: Size error: Distributions must all be of length n (number of states).
```
IMO the reason is that processing of the discrete trajectories internally misses out about observable states that are a) not in the discrete trajectory and b) have state indices larger than the max state index in the given trajectories.

**Comment by thempel** [comment]
Could be related to this line? https://github.com/deeptime-ml/deeptime/blob/f6217a6493f475a491023814f71d18f34e3689af/deeptime/markov/_bindings/include/discrete_trajectories.h#L38

**Comment by thempel** [comment]
fixed by #150
